proposals/secrets-support.md

```diff
@@ -0,0 +1,85 @@
+# First class secrets support in devcontainers CLI
+
+## What are secrets
+Secrets are variables that hold sensitive values and needs to be handled securely at all times (API keys, passwords etc.). Users can change a secret's value at anytime, and the devcontainers CLI should support dynamically changing secrets without having to rebuild the container.
+
+## Goal
+
+Support secrets as a first class feature in devcontainers CLI.
+
+This feature should comprise of the ability to securely pass in the secrets in CLI commands, make them available for users to consume similar to remoteEnv and containerEnv, and be processed and handled securely at all times.
+This functionality will allow consumers to be able to use secrets more predictably and securely with devcontainers CLI.
+
+## Motivation
+
+Today Codespaces pass secret variables as `remoteEnv` since there are no other mechanisms to pass them explicitly as secrets. The devcontainers CLI do not (and need not) treat `remoteEnv` as secrets. Having explicit support for secrets will help eliminating security threats such as leaking secrets, as well as in improving customer confidence.
+
+## Proposal
+
+A [supporting tool](https://containers.dev/supporting#tools) (ie, the [dev container CLI reference implementation](https://github.com/devcontainers/cli) should behave according to the following properties:
+
+  1. Ability to pass secrets to commands
+  2. Apply/use secrets similar to `remoteEnv`
+  3. Securely handle secrets
+
+### Passing secrets to commands
+Secrets are not part of dev containers specification and we do not expect users to store secrets inside devcontainer.json file.
+
+Secrets should be passed to the following devcontainers CLI commands.
+
+#### **Phase 1**
+ - up
+ - run-user-commands
+
+`up`, `run-user-commands` will use the secrets for usecases like lifecycle hooks (similar to remoteEnv). We do not plan secrets support for `build` command since we do not have a demonstrated requirement, as well as to avoid potentially baking in the user secrets into the docker images.
+
+#### **Phase 2**
+In phase 2 the support for secrets can be expanded to additional commands such as `set-up` and `exec`.
+
+#### **Input format**
+Best way to pass secrets would be through a json/env file, where in the file path is supplied in a command line argument `--secrets-file`.
```

Comment:
Idea is that user can keep the secrets in a file (and store it anywhere on the machine, that they feel safe). It need not be inside devcontainer directory (and of course we don't expect them to check in this file.)
But to your point many of the secrets may be project specific and they could be stored along side other .devcontainer files. But I am not certain if we want to default to that.

## Future enhancements
Secondly, in the future we can expand to support secrets from additional sources such as Windows credential manager, Mac key chains, Azure keyvault etc. in addition to the secrets file support.
I think, to be flexible we might not want to limit the source to be the project dir.

Reply:
Sill wondering how we are going to make it usable with the Dev Containers extension. Codespaces will get the secrets from GitHub's configuration for the repository. Will the Dev Containers extension have to introduce its own config file to associate secret files with devcontainer.json files?

Since each implementation can introduce its own solution for looking up secrets, the Dev Containers extension could look for a secrets file next to the devcontainer.json file without that location being part of the specification. But maybe this is an indication that we should add the default location to the OSS CLI? Otherwise the OSS CLI - when used locally - will not behave the same way the Dev Containers extension does.

Reply:
I have removed the CLI specific notes from this spec to keep it generic.

All CLI specific notes and recommendations has been moved to [an issue in CLI repo,](https://github.com/devcontainers/cli/issues/507) for consideration at the implementation time